### ISSUE TITLE
Trouble using xdebug with Netbeans & v0.13.0-beta.2

### ISSUE THREAD

**Issue by rbertolette** [kalabox/kalabox#1491]
Feature/Enhancement Request
===========================
I'd like advice on how to setup Netbeans to work with xdebug and kbox v0.13.0-beta.2

**setup:**
osX 10.10.5, MacBook Pro (Retina, 15-inch, Late 2013), 2.3 GHz Intel Core i7, 16 GB 1600 MHz DDR3

**Netbeans config:**
Product Version: NetBeans IDE 8.1 (Build 201510222201)
Updates: NetBeans IDE is updated to version NetBeans 8.1 Patch 1
Java: 1.8.0_60; Java HotSpot(TM) 64-Bit Server VM 25.60-b23
Runtime: Java(TM) SE Runtime Environment 1.8.0_60-b27
System: Mac OS X version 10.10.5 running on x86_64; UTF-8; en_US (nb)
User directory: /Users/Reed/Library/Application Support/NetBeans/8.1
Cache directory: /Users/Reed/Library/Caches/NetBeans/8.1

I've been using 0.12.0-beta.1 successfully with Netbeans & xdebug

As a Kalabox user who prefers Netbeans with Pantheon created kbox, I'd like to upgrade to the latest version of kbox and still be able to debug using Netbeans.

With kbox 0.12.0-beta.1 I'm able to use Netbeans by adding the following to line to the <kbox-root>/config/php/php.ini

`xdebug.remote_port=9001`
NOTE: this is probably only necessary because I use the following settings (Netbeans->Preferences):
![netbeansprefsphpdebug](https://cloud.githubusercontent.com/assets/50540/17115394/d4a07bd6-524e-11e6-926b-dc431249e711.png)

From phpinfo() I can see xdebug v2.2.5 is loading.

Right clicking on my Netbeans project and selecting Properties, Run Configuration category, clicking advanced I've got the following settings:
![netbeansadvwebconfig](https://cloud.githubusercontent.com/assets/50540/17116158/34e0f6ee-5252-11e6-8116-c05ff7bd7c73.png)

I've set the path by double-clicking in the Path Mapping table under the server path column and typing:
`/code/`
then clicking the "..." icon to the right of the row and selecting the code directory

(Hopefully this will aid other Netbeans users)

However, when I upgraded to kalabox-v0.13.0-beta.2.dmg, I could no longer get xdebug to work consistently. I was able to get it to start a debug session if I selected "Stop at First Line" from the Netbeans->Preferences->PHP->Debug screen (see first grab above.) It stops at `<kbox root>/config/php/prepend.php`
I'm able to step through, but neither inserted breakpoints don't work, nor xdebug_break() work.

**Comment by pirog** [comment]
@rbertolette i'm not as familiar with xdebug on NetBeans but i've definitely been able to use xdebug pretty successfully with sublimetext. would definitely like to get this sorted out and then added to our documentation so we can help netbeans users. 

So the obvious things that jump out to me just reading your issues are 

1. Why are you using port 9001 instead of the default 9000?
2. Are you using an app that you created BEFORE you upgraded to the latest version of Kalabox or did you recreate the app after the upgrade? 
3. Are you sure your local project path is actually for the app you are using? ie did you happen to install two versions of the same app but are using a codebase that doesnt actually map to the url. 
4. Can you get breakpoints to trigger in the prepend.php at all?

**Comment by rbertolette** [comment]
@pirog I appreciate your quick reply. I'm using the reinstalled stable version of  0.12.0-beta.1, so I can't actually try anything new at the moment (question 4), but I can answer the other questions.

_This is admittedly kinda a mess, so I'm not convinced you'll be able to see this as a typical setup._

**1. Why are you using port 9001 instead of the default 9000?**
In 2010 I created local programming environments by switching between php version environments that I'd installed with MacPorts.

In 2014 I switched to homebrew, & completely removed MacPorts. Details [here](https://stackoverflow.com/questions/2963027/netbeans-xdebug-php-not-working/29573712#29573712 ).  BTW, I'm not using all the setting in ext-xdebug.ini  from that link. I found that I only needed to add `xdebug.remote_port=9001` as mentioned above. 

You may notice that this approach:

> installs DNSMasq which runs on xdebug's default port (9000), so you'll need to change the xdebug port to something else (9001 works fine.)

NOTE: When I'm using kbox, I run a script that completely turns off my local php & apache environments:

```
case "$1" in

'off')
launchctl unload -Fw ~/Library/LaunchAgents/homebrew.mxcl.php53.plist
launchctl unload -Fw ~/Library/LaunchAgents/homebrew.mxcl.php55.plist
brew unlink php53
brew unlink php55
sleep 2
sudo apachectl stop
;;
(etc)

esac
```

I left the port setting a 9001 so that I didn't have to change it every time I switch to a kbox environment.

**2. Are you using an app that you created BEFORE you upgraded to the latest version of Kalabox or did you recreate the app after the upgrade?**

BLUF Yes, I'm pretty sure.

DETAILS:
The app I created originally w/ 0.12.0-beta.1 was located in the ~/.kalabox/apps directory. I made a timemachine bu, &  copied the ~/.kalabox/apps/<my_app> directory to a different volume, & changed the name before uninstalling 0.12.0-beta.1 (w/ the ininstall.sh script on the 0.12.0-beta.1.dmg .) At this point, I hadn't realized how to set the Server Path. Breakpoints added w/ NetBeans didn't work, however I was able to use `xdebug_break();` in the code, & use the Call Stack to figure out where I was. Variables were populated as usual.

Then I installed 0.13.0-beta.2 and used the cli to create the new Pantheon app (with --nofiles & --nodb) downloaded the files & database from Pantheon & loaded the database manually using mysql from the command line. This the new app created the app root directly in my ~/ directory. At this point I reread some of the docs that showed setups for xdebug & Sublime & figured out what the correct settings for the Server Path would be. Still, this version that would only stop with Stop at First Line setting in Netbeans.

I uninstalled 0.13.0-beta.2, reinstalled 0.12.0-beta.1, ran kbox create pantheon using --nofiles & --nodb again. (I don't think I'd deleted the 0.13.0-beta.2 app from my ~/ directory. Now the app was still located in the ~/ directory. And it appeared to keep my .git changes. I recreated the Project in NetBeans, set the Server Path, and discovered that the breakpoints I set in the NetBeans UI were now working (I no longer needed to add xdebug_break() code to break.)

I'm assuming that If I'd set the Server Path correctly in the original 0.12.0-beta setup, it would have worked as well.

**3. Are you sure your local project path is actually for the app you are using? ie did you happen to install two versions of the same app but are using a codebase that doesnt actually map to the url.**

When I was trying to use 0.13.0-beta.2, I'm pretty sure there was only one version of the app installed.

**4. Can you get breakpoints to trigger in the prepend.php at all?**
I appreciate your attention to this issue, but I can't now reinstall 0.13.0-beta.2 to test this out. Possibly I'll have some time later this week to test this out. I'll let you know then.

**Comment by rbertolette** [comment]
@pirog,
Just an afterthought, I saved my phpinfo() screen from the  0.13.0-beta.2. It shows that the version of xdebug running was Xdebug v2.2.7, yet my current version phpinfo() running 0.12.0-beta shows Xdebug v2.2.5 running. It's showing it `/root/.phpbrew/php/php-5.3.29/var/db/xdebug.ini` in the Additional .ini files parsed section. I'm unclear as to where that path lead, I tried to find it w/ 
`
find / -name xdebug.ini
`
with no luck, I think I'm unclear on the docker paradigm.
